solutions/gold/usaco-1235.mdx

```diff
@@ -0,0 +1,175 @@
+---
+id: usaco-1235
+source: USACO Gold 2022 Open
+title: Balancing a Tree
+author: Daniel Zhu
+---
+
+<Spoiler title="Hint">
+
+Consider binary searching on the minimum imbalance. For a given imbalance $k$, how can we check whether a valid tree with imbalance $\leq k$ exists?
+
+</Spoiler>
+
+<Spoiler title="Solution Part 1">
+
+We can try to construct the tree bottom-up, ensuring as we move up the tree that the value of each node is still within $k$ of the values of its descendants. Thus, we can store for each node the range of values it's allowed to take on: for a leaf $i$, this will simply be $[l_i, r_i]$, but this restricts its parent (and all ancestors) $p_i$ to the range $[l_i - k, r_i + k]$, in addition to $p_i$'s own restriction of $[l_{p_i}, r_{p_i}]$. If at any point a node is restricted to a degenerate range (i.e. $l > r$), we know a tree with imbalance $k$ cannot exist.
+
+Here's the snippet for this checker function:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+/**
+ * computes restrictions on each node in tree
+ * @return all restrictions if valid, otherwise empty array
+ */
+vector<pair<int, int>> check(int k) {
+	// L[i], R[i] -> original restrictions (from input) on i
+	vector<pair<int, int>> lr(n + 1);
+	for (int i = 1; i <= n; i++) {
+		// all ancestors of i are restricted to [l_i - k, r_i + k]
+		lr[i] = {L[i] - k, R[i] + k};
+	}
+
+	// this works because p_i < i
+	for (int i = n; i > 0; i--) {
+		// check for degenerate ranges
+		if (lr[i].first > lr[i].second) { return {}; }
+		if (p[i] && (lr[i].first > R[p[i]] || lr[i].second < L[p[i]])) {
+			return {};
+		}
+		// propagate restrictions to parent
+		lr[p[i]] = {
+			max(lr[p[i]].first, lr[i].first),
+			min(lr[p[i]].second, lr[i].second)
+		}
+	}
+	// return a vector of all restrictions
+	// (lr[i] -> range to which node i is restricted)
+	return lr;
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+</Spoiler>
+
+<Spoiler title="Solution Part 2">
```

Comment:
could you make one big spoiler for solution

and then subspoilers for like part1, part2, ...

just to keep things consistent :)

Reply:
okok gotchu